File: Linked Lists/ll.cpp
//Singly LL

template <typename T>
class LinkedList {
private:

    struct Node {
        T data;
        Node* next;

        Node(const T& value) {
            data = value;
            next = nullptr;
        }
    };

    Node* head;
    int sz;

public:

    // Constructor
    LinkedList() {
        head = nullptr;
        sz = 0;
    }

    // Destructor
    ~LinkedList() {
        Node* curr = head;

        while (curr != nullptr) {
            Node* temp = curr;
            curr = curr->next;
            delete temp;
        }
    }

    // Push Front
    void push_front(const T& value) {
        Node* newNode = new Node(value);

        newNode->next = head;
        head = newNode;

        sz++;
    }

    // Push Back
    void push_back(const T& value) {
        Node* newNode = new Node(value);

        if (head == nullptr) {
            head = newNode;
            sz++;
            return;
        }

        Node* curr = head;

        while (curr->next != nullptr) {
            curr = curr->next;
        }

        curr->next = newNode;

        sz++;
    }

    // Pop Front
    void pop_front() {
        if (head == nullptr) {
            throw std::out_of_range("List is empty");
        }

        Node* temp = head;

        head = head->next;

        delete temp;

        sz--;
    }

    // Size
    int size() {
        return sz;
    }

    // Empty
    bool empty() {
        return sz == 0;
    }
};


//Doubly LL
template <typename T>
class DoublyLinkedList {
private:
    struct Node {
        T data;
        Node* prev;
        Node* next;

        Node(const T& value) {
            data = value;
            prev = nullptr;
            next = nullptr;
        }
    };

    Node* head;
    Node* tail;
    int sz;

public:
    DoublyLinkedList() {
        head = nullptr;
        tail = nullptr;
        sz = 0;
    }

    ~DoublyLinkedList() {
        Node* curr = head;

        while (curr != nullptr) {
            Node* temp = curr;
            curr = curr->next;
            delete temp;
        }
    }

    void push_front(const T& value) {
        Node* newNode = new Node(value);

        if (head == nullptr) {
            head = tail = newNode;
        } else {
            newNode->next = head;
            head->prev = newNode;
            head = newNode;
        }

        sz++;
    }

    void push_back(const T& value) {
        Node* newNode = new Node(value);

        if (tail == nullptr) {
            head = tail = newNode;
        } else {
            tail->next = newNode;
            newNode->prev = tail;
            tail = newNode;
        }

        sz++;
    }

    void pop_front() {
        if (head == nullptr) {
            throw std::out_of_range("List is empty");
        }

        Node* temp = head;

        if (head == tail) {
            head = tail = nullptr;
        } else {
            head = head->next;
            head->prev = nullptr;
        }

        delete temp;
        sz--;
    }

    void pop_back() {
        if (tail == nullptr) {
            throw std::out_of_range("List is empty");
        }

        Node* temp = tail;

        if (head == tail) {
            head = tail = nullptr;
        } else {
            tail = tail->prev;
            tail->next = nullptr;
        }

        delete temp;
        sz--;
    }

    int size() const {
        return sz;
    }

    bool empty() const {
        return sz == 0;
    }
};

File: Linked Lists/ll.py
#Singly LL
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:

    def __init__(self):
        self.head = None
        self.size = 0

    # Push Front
    def push_front(self, value):

        new_node = Node(value)

        new_node.next = self.head

        self.head = new_node

        self.size += 1

    # Push Back
    def push_back(self, value):

        new_node = Node(value)

        if self.head is None:
            self.head = new_node
            self.size += 1
            return

        curr = self.head

        while curr.next:
            curr = curr.next

        curr.next = new_node

        self.size += 1

    # Pop Front
    def pop_front(self):

        if self.head is None:
            raise Exception("List is empty")

        self.head = self.head.next

        self.size -= 1


#Doubly LL
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def push_front(self, value):
        new_node = Node(value)

        if self.head is None:
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

        self.size += 1

    def push_back(self, value):
        new_node = Node(value)

        if self.tail is None:
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node

        self.size += 1

    def pop_front(self):
        if self.head is None:
            raise Exception("List is empty")

        value = self.head.data

        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None

        self.size -= 1
        return value

    def pop_back(self):
        if self.tail is None:
            raise Exception("List is empty")

        value = self.tail.data

        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None

        self.size -= 1
        return value